isabelle: fix permissions of file copied from store

LogicalOverflow opened the pull request:
## Description of changes

Isabelle copies files from the store and then edits them when building documents. Because the copied files keep the read-only permissions from the store, this results in errors. To work around the issues, this commit adds a patch that simply marks all files copied as writable by the owner.

Fixes #289529 (which contains more details)

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

drupol commented on pkgs/applications/science/logic/isabelle/default.nix:
```diff
@@ -70,6 +70,8 @@ in stdenv.mkDerivation (finalAttrs: rec {
         hash = "sha256-Tzxxs0gKw6vymbaXIzH8tK5VgUrpOIp9vcWQ/zxnRCc=";
       };
 
+  patches = [ ./fix-copied-permissions.patch ];
```
Each patch in nixpkgs applied to the upstream source should be documented. Out-of-tree patches, fetched using `fetchpatch`, are preferred.

In each patch comment, please explain the purpose of the patch and link to the relevant upstream issue if possible. If the patch has been merged upstream but is not yet part of the released version, please note the version number or date in the comment such that a future maintainer updating the nix expression will know whether the patch has been incorporated upstream and can thus be removed from nixpkgs.

Furthermore, please use a _stable_ URL for the patch. Rather than, for example, linking to a GitHub pull request of the form `https://github.com/owner/repo/pull/pr_number.patch`, which would change every time a commit is added or the PR is force-pushed, link to a specific commit patch in the form `https://github.com/owner/repo/commit/sha.patch`.

Here are two good examples of patch comments:

```nix
mkDerivation {
  …

  patches = [
    # Ensure RStudio compiles against R 4.0.0.
    # Should be removed next 1.2.X RStudio update or possibly 1.3.X.
    (fetchpatch {
      url = "https://github.com/rstudio/rstudio/commit/3fb2397c2f208bb8ace0bbaf269481ccb96b5b20.patch";
      sha256 = "0qpgjy6aash0fc0xbns42cwpj3nsw49nkbzwyq8az01xwg81g0f3";
    })
  ];
}
```

```nix
mkDerivation {
  …

  patches = [
    # Allow building with bison 3.7
    # PR at https://github.com/GoldenCheetah/GoldenCheetah/pull/3590
    (fetchpatch {
      url = "https://github.com/GoldenCheetah/GoldenCheetah/commit/e1f42f8b3340eb4695ad73be764332e75b7bce90.patch";
      sha256 = "1h0y9vfji5jngqcpzxna5nnawxs77i1lrj44w8a72j0ah0sznivb";
    })
  ];
}
```

LogicalOverflow replied:
Thanks for the pointer, I added a comment explaining the patch and linked the relevant nixpkgs issue.

I don't think the patch makes sense to include upstream, as this feels very much like a nix-specific issue. (Copying files from a template including file permissions feels like a sane approach in general, and it's only an issue because nix makes the template read-only.)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3802

drupol reviewed (changes requested):
Looks like we should update `meta.platforms`.

LogicalOverflow commented:
@drupol I am a bit lost here; to what should `meta.platforms` be changed? And why?

If it is because of OfBorg skipping/failing the darwin build, that seems to be happening for some time now (at least the OfBorg runs on the recent commits to Isabelle seem to do that too), and Hydra successfully builds Isabelle for dawin anyways.

LogicalOverflow committed:
isabelle: fix permissions of file copied from store

Isabelle copies files from the store and then edits them when building
documents. Because the copied files keep the read-only permissions from the
store, this results in errors. To work around the issues, this commit adds a
patch that simply marks all files copied as writable by the owner.

ShamrockLee commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 289575`
Commit: [`6f4f359b6354cbb87d908aa62e32bfbcf7418649`](https://github.com/NixOS/nixpkgs/commit/6f4f359b6354cbb87d908aa62e32bfbcf7418649) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/6f4f359b6354cbb87d908aa62e32bfbcf7418649..pull/289575/head))
Merge: [`50c211ebe13a1c04633223c4aca71d20b27a6115`](https://github.com/NixOS/nixpkgs/commit/50c211ebe13a1c04633223c4aca71d20b27a6115)

Logs: https://github.com/ShamrockLee/nixpkgs-review-gha/actions/runs/19116405421


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>tlaps</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>tlaps</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>tlaps</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>isabelle</summary>
<pre>@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;buildPhase&quot; }
Building isabelle setup
Building HOL heap
### Building Isabelle/Scala (/nix/build/nix-10152-1213594871/Isabelle2025/lib/classes/isabelle.jar) ... (sandbox = relaxed)
### Building Demo (/nix/build/nix-10152-1213594871/Isabelle2025/src/Tools/Demo/lib/demo.jar) ... (sandbox = relaxed)
Started at Wed Nov 5 21:37:22 GMT 2025 (polyml-for-isabelle-2025_x86_64-darwin on sjc20-cw713-bc0ea16a-04dd-414d-a002-498d6454c0e4-8EC0AAA1B49C)
ISABELLE_TOOL_JAVA_OPTIONS=&quot;-Djava.awt.headless=true -Xms512m -Xmx4g -Xss16m&quot;
ISABELLE_BUILD_OPTIONS=&quot;&quot;

ML_PLATFORM=&quot;x86_64-darwin&quot;
ML_HOME=&quot;/nix/store/3qw0ykvbkpz4b0k3ln7k293rgp1ckr7v-polyml-for-isabelle-2025/bin&quot;
ML_SYSTEM=&quot;polyml-for-isabelle-2025&quot;
ML_OPTIONS=&quot;--minheap 1000&quot;

2025-11-05 21:37:26.235 java[18230:82974] CoreText: System LastResort not available, using built-in copy.
Session Pure/Pure
Session Misc/Tools
Session HOL/HOL (main)
Building Pure ...
*** I/O error: Operation not permitted</pre>
</details>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>tlaps</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>isabelle</summary>
<pre>@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;buildPhase&quot; }
Building isabelle setup
Building HOL heap
### Building Isabelle/Scala (/nix/build/nix-7560-2003511708/Isabelle2025/lib/classes/isabelle.jar) ... (sandbox = relaxed)
### Building Demo (/nix/build/nix-7560-2003511708/Isabelle2025/src/Tools/Demo/lib/demo.jar) ... (sandbox = relaxed)
Started at Wed Nov 5 21:18:56 GMT 2025 (polyml-for-isabelle-2025_aarch64-darwin on iad20-gt1020-5ebae457-79bd-4250-99c0-fb3b868b5419-12673C50555C)
ISABELLE_TOOL_JAVA_OPTIONS=&quot;-Djava.awt.headless=true -Xms512m -Xmx4g -Xss16m&quot;
ISABELLE_BUILD_OPTIONS=&quot;&quot;

ML_PLATFORM=&quot;aarch64-darwin&quot;
ML_HOME=&quot;/nix/store/xkrj1sdigwx4lyf51whirqy7m2l9k0x3-polyml-for-isabelle-2025/bin&quot;
ML_SYSTEM=&quot;polyml-for-isabelle-2025&quot;
ML_OPTIONS=&quot;--minheap 1000&quot;

2025-11-05 21:18:56.351 java[10001:48961] CoreText: System LastResort not available, using built-in copy.
Session Pure/Pure
Session Misc/Tools
Session HOL/HOL (main)
Building Pure ...
*** I/O error: Operation not permitted</pre>
</details>
</details>
